lib/compile/index.js

```diff
@@ -0,0 +1,96 @@
+import most from "most";
+import escodegen from "escodegen";
+
+import applyPlugins from "./apply-plugins";
+import construct from "./construct";
+import compileModules from "./modules/compile";
+import bootstrapBundles from "./bundles/bootstrap";
+import hashBundle from "./bundles/hash";
+import interpolateFilename from "./bundles/interpolate-filename";
+import dedupeExplicit from "./bundles/dedupe-explicit";
+import dedupeImplicit from "./bundles/dedupe-implicit";
+
+export default function compile (opts, compilation) {
+  compilation = compilation || {
+    cache: {
+      modulesByAbsPath: Object.create(null)
+    },
+    applyPlugins: (ev, ...args) => applyPlugins(opts.plugins, ev, ...args),
```

Comment:
I share most of your objections to the plugin architecture, and this is one of the things I want to completely rethink.  To answer this question, tho, `applyPlugins` was only intended to be used internally.